content/blog/2022-05-23-vectorization/index.md

```diff
@@ -0,0 +1,238 @@
+
+I also track the number of vectors created statically. These results are perhaps more revealing of the ability of the vectorization pass to find opportunities to create vector instructions. 
+
+<p align="center">
+<img src="static_vector_creation.png">
+title="Vectors Statically Created" style="zoom:25%;"
+</p>
+
+Most of the benchmarks have no vector instructions created statically, but for those that do, between 1 and 4 vector instructions are created. 
+
+Finally, no benchmark actually had code that utilized `vecmove`, indicating that sharing and reusing of vectors is much more complex to do, compared to the basic schemes I devised. Only `pythagorean-triple` reused a vector.
+
+## Shortcomings in the Vectorization Schemes
+
+There were several major issues with the vectorization schemes that I proposed. First, I did not end up creating vector instructions like `vecmac`. However, some benchmarks using matrices used multiple sequences of consecutive `mul` and `add` instructions. In particular, benchmarks like `mat-mul`, `mat-inv`, and `cholesky` fell into this category, and would be targets for creating `vecmac` instructions. 
+
+Next, the preprocessing passes are not as strong as I intended them to be. For instance, the unrolling pass is very simple in how it detects whether a loop can be unrolled. In fact, the unrolling pass never actually runs, because the only benchmark it could have run on, `mat-mul`, did not hardcode the dimensions of the matrix, instead passing it in as a parameter. As a result, I would need to improve the unrolling pass to make it detect a wider range of fully-unrollable loops. In addition, the alias analysis pass was designed to be very basic, generally only being able to disambiguate between two different arrays. A major weakness is that the alias analysis pass only assigns one address to an entire array, failing to disambiguate between different locations within an array. This means that some instances of store movements cannot occur successfully, because the alias analysis will report two addresses in an array can alias, when in fact the two addresses are at different memory locations. Hence, it would probably be a good idea to augment the alias analysis to model each cell in the array.
```

Comment:
I think this is emblematic of a broader lesson about auto-vectorization in compilers in general. It depends so heavily on the success of "standard" compiler analyses—*especially* alias analysis! Really neat to see this observation play out in practice here.